ros_gz_bridge/src/convert/marine_acoustic_msgs.cpp

```diff
@@ -0,0 +1,165 @@
+      gz::msgs::DVLKinematicEstimate::DVL_REFERENCE_SHIP);
+
+    auto beam_unit = ros_msg.beam_unit_vec[i];
+    auto beam_velocity = ros_msg.beam_velocity[i];
+    beam->mutable_velocity()->mutable_mean()->set_x(beam_unit.x * beam_velocity);
+    beam->mutable_velocity()->mutable_mean()->set_y(beam_unit.y * beam_velocity);
+    beam->mutable_velocity()->mutable_mean()->set_z(beam_unit.z * beam_velocity);
+
+    // Unsupported.
+    for (auto j = 0; j < 9; ++j) {
+      beam->mutable_velocity()->add_covariance(0);
+    }
+
+    beam->mutable_range()->set_mean(ros_msg.range[i]);
+    beam->mutable_range()->set_variance(ros_msg.range_covar[i]);
+
+    // rssi not supported in ROS.
+    // nsd not supported in ROS.
+
+    beam->set_locked(true);
+  }
+
+  // Velocity.
+  gz_msg.mutable_velocity()->set_reference(
+    gz::msgs::DVLKinematicEstimate::DVL_REFERENCE_SHIP);
+  convert_ros_to_gz(ros_msg.velocity, (*gz_msg.mutable_velocity()->mutable_mean()));
+  for (auto i = 0; i < 9; ++i) {
+    gz_msg.mutable_velocity()->add_covariance(ros_msg.velocity_covar[i]);
+  }
+
+  // Target.
+  if (ros_msg.velocity_mode == marine_acoustic_msgs::msg::Dvl::DVL_MODE_BOTTOM) {
+    gz_msg.mutable_target()->set_type(gz::msgs::DVLTrackingTarget::DVL_TARGET_BOTTOM);
+  } else if (ros_msg.velocity_mode == marine_acoustic_msgs::msg::Dvl::DVL_MODE_WATER) {
+    gz_msg.mutable_target()->set_type(gz::msgs::DVLTrackingTarget::DVL_TARGET_WATER_MASS);
+  } else {
+    gz_msg.mutable_target()->set_type(gz::msgs::DVLTrackingTarget::DVL_TARGET_UNSPECIFIED);
+  }
+
+  // Range and position Unavailable.
+
+  gz_msg.set_status(0);
+}
+
+template<>
+void
+convert_gz_to_ros(
+  const gz::msgs::DVLVelocityTracking & gz_msg,
+  marine_acoustic_msgs::msg::Dvl & ros_msg)
+{
+  convert_gz_to_ros(gz_msg.header(), ros_msg.header);
+
+  if (gz_msg.target().type() == gz::msgs::DVLTrackingTarget::DVL_TARGET_BOTTOM) {
+    ros_msg.velocity_mode = marine_acoustic_msgs::msg::Dvl::DVL_MODE_BOTTOM;
+  } else if (gz_msg.target().type() == gz::msgs::DVLTrackingTarget::DVL_TARGET_WATER_MASS) {
+    ros_msg.velocity_mode = marine_acoustic_msgs::msg::Dvl::DVL_MODE_WATER;
+  }
+
+  if (gz_msg.type() == gz::msgs::DVLVelocityTracking::DVL_TYPE_PISTON) {
+    ros_msg.dvl_type = marine_acoustic_msgs::msg::Dvl::DVL_TYPE_PISTON;
+  } else if (gz_msg.type() == gz::msgs::DVLVelocityTracking::DVL_TYPE_PHASED_ARRAY) {
+    ros_msg.dvl_type = marine_acoustic_msgs::msg::Dvl::DVL_TYPE_PHASED_ARRAY;
+  }
+
+  convert_gz_to_ros(gz_msg.velocity().mean(), ros_msg.velocity);
+
+  for (auto i = 0; i < 9; ++i) {
+    if (gz_msg.velocity().covariance_size() > i) {
+      ros_msg.velocity_covar[i] = gz_msg.velocity().covariance()[i];
+    } else {
+      ros_msg.velocity_covar[i] = -1;
+    }
+  }
+
+  ros_msg.altitude = -1;
+  ros_msg.course_gnd = std::atan2(ros_msg.velocity.y, ros_msg.velocity.x);
+  ros_msg.speed_gnd = std::sqrt(ros_msg.velocity.x * ros_msg.velocity.x + ros_msg.velocity.y *
+      ros_msg.velocity.y);
```

Comment:
@caguero meta: I would double check frames are correct. It's not clear to me what the ROS message definition means by ground speed.